test_backend.py

```diff
@@ -117,13 +123,201 @@ def test_robot_is_repaired(damages_before, tile, damages_after):
+                        [((0, 0), (0, 0)),
+                         ((1, 0), (1, 0)),
+                         ((2, 0), (2, 1)),
+                         ((3, 0), (3, 1)),
+                         ])
+def test_robot_is_stopped_by_wall(input_coordinates, output_coordinates):
+    """
+    Take robot's coordinates, move's direction and distance and assert robot
+    was moved to correct coordinates.
+    A special map test_walls was created in order to test this feature.
+    """
+    board = get_board("maps/test_walls.json")
+    robot = Robot(None, None, None, input_coordinates)
+    state = State(board, [robot], 25)
+    robot.move(Direction.N, 2, state)
+    assert robot.coordinates == output_coordinates
+
+
+# LaserTile
+
+@pytest.mark.parametrize(("input_coordinates", "add_damages"),
+                        [((1, 2), 0),
+                         ((3, 1), 0),
+                         ((2, 1), 0),
+                         ((0, 1), 2),
+                         ((3, 3), 4),
+                         ((2, 3), 3),
+                         ])
+def test_robot_is_damaged_by_laser(input_coordinates, add_damages):
+    """
+    When robot stands on laser tile, he is damaged according to the laser strength, but only if there is no obstacle in the way.
+    If there are obstacles, damage count changes accordingly.
+    A special map test_laser was created in order to test this feature.
+    """
+    board = get_board("maps/test_laser.json")
+    robot_obstacle1 = Robot(Direction.N, None, None, (1, 1))
+    robot_obstacle2 = Robot(Direction.N, None, None, (3, 2))
+    robot = Robot(Direction.E, None, None, input_coordinates)
+
+    # Get count of current robot's damage (as we change it in backend for
+    # development purposes). When game is finished,
+    # the math behind counting final damages can be removed.
+    damages_before = robot.damages
+    state = State(board, [robot_obstacle1, robot_obstacle2, robot], 16)
+    apply_tile_effects(state)
+
+    # Assertion that the sum of initial damages and added damages is as expected
+    # We cannot determine the precise number at this point as our robots
+    # are initialized with diferent count of damages (for devel purposes).
+    assert robot.damages == damages_before + add_damages
+
+
+# PusherTile
+
+@pytest.mark.parametrize(("game_round", "tile", "output_coordinates"),
+                        [(1, PusherTile(Direction.N, None, [{'value': 1}]), (1, 0)),
+                         (2, PusherTile(Direction.N, None, [{'value': 1}]), (1, 1)),
+                         (3, PusherTile(Direction.N, None, [{'value': 0}]), (1, 1)),
+                         (4, PusherTile(Direction.N, None, [{'value': 0}]), (1, 0)),
+                         (5, PusherTile(Direction.N, None, [{'value': 1}]), (1, 0)),
+                         ])
+def test_robot_is_pushed_at_the_correct_round(game_round, tile, output_coordinates):
+    """
+    When robot is standing on a PusherTile, he should be pushed against it.
```

Comment:
Ta druhá část věty mi moc nedává smysl. Robot je posunutý proti něčemu?

Reply:
"In the opposite direction"? Moc nevím, jak to vyjádřit, že ho pusher vlastně posune dozadu ve směru pusheru.

Reply:
Mne přijde, že tak je to OK. Pusher na východní stěně posouvá robota na západ. Sedí mi to "selským" rozumem. Pokusím se napsat to vysvětlení trošku líp.